### DISCUSSION TITLE
RuntimeError when attempting to run an already setup cell

### DISCUSSION THREAD

**Discussion by lungsi** [arbor-sim/arbor#1204] [Q&A] [answered]
Arbor Version: 0.4.1.dev0

# 1. RuntimeError Type 1
Running my cable cell `cell_model.run(100)` returns a RuntimeError
```
---------------------------------------------------------------------------
RuntimeError                              Traceback (most recent call last)
<ipython-input-6-41323ba3c1b4> in <module>
----> 1 cell_model.run(100) # 100ms

RuntimeError: no mechanism passive in catalogue
```

The `mechanisms` and their decorations on the cell are as follows
```
# Set mechanisms
leaky_chnl = arbor.mechanism('passive/el=-54.3', {'g': 0.0003})
hh_na_k_chnl = arbor.mechanism("hh", {"ena": 50., "gnabar": 0.120,
                                      "ek": -77., "gkbar": 0.036,
                                      "gl": 0.0003})
# Attach mechanism
cell.paint('"soma"', hh_na_k_chnl)
cell.paint('"axon"', hh_na_k_chnl)
cell.paint('"dend"', hh_na_k_chnl)
cell.paint('"everywhere"', leaky_chnl)
```

**I get the error (RuntimeError above) only when I try to run the model: I don't get any error during the setup.**

# 2. RuntimeError Type 2
If try to run the same model but **without** `cell.paint('"everywhere"', leaky_chnl)` I get
```
---------------------------------------------------------------------------
RuntimeError                              Traceback (most recent call last)
<ipython-input-6-41323ba3c1b4> in <module>
----> 1 cell_model.run(100) # 100ms

RuntimeError: mechanism hh has no parameter ena
```

# 3. RuntimeError Type 3
If try to run the same model but **without** `cell.paint('"soma"', hh_na_k_chnl); cell.paint('"axon"', hh_na_k_chnl); cell.paint('"dend"', hh_na_k_chnl)` I get
```
---------------------------------------------------------------------------
RuntimeError                              Traceback (most recent call last)
<ipython-input-6-41323ba3c1b4> in <module>
----> 1 cell_model.run(100) # 100ms

RuntimeError: no mechanism passive in catalogue
```

## How can I fix the cause for all the three errors?

**Comment by noraabiakar** [comment]
Hi, 

The passive mechanism is called `pas` in arbor's default mechanism catalogue. If you use `pas` that should fix the first and third errors. 
As for the second error: in arbor's version of the `hh` mechanism, `ena` and `ek` are not an `assigned` parameters. They refer to the reversal potentials of the sodium channel and potassium channel respectively; and you can paint the initial reversal potentials of ions (as well as their initial concentrations) directly on the cell, similar to how you painted mechanisms: 
```
# Set mechanisms
leaky_chnl = arbor.mechanism('pas/e=-54.3', {'g': 0.0003})
hh_na_k_chnl = arbor.mechanism("hh", {"gl": 0.0003})

# Attach mechanism and paint ion parameters
cell.paint('"soma"', "na", rev_pot=50)
cell.paint('"soma"', "k", rev_pot=-77)
cell.paint('"soma"', hh_na_k_chnl)

cell.paint('"axon"', "na", rev_pot=50)
cell.paint('"axon"', "k", rev_pot=-77)
cell.paint('"axon"', hh_na_k_chnl)

cell.paint('"dend"', "na", rev_pot=50)
cell.paint('"dend"', "k", rev_pot=-77)
cell.paint('"dend"', hh_na_k_chnl)

cell.paint('"everywhere"', leaky_chnl)
```
You can also `join` regions to reduce the lines of code: 

```
leaky_chnl = arbor.mechanism('pas/el=-54.3', {'g': 0.0003})
hh_na_k_chnl = arbor.mechanism("hh", {"gl": 0.0003})

# Attach mechanism and paint ion parameters
cell.paint('(join (region "soma") (region "dend") (region "axon"))', "na", rev_pot=50)
cell.paint('(join (region "soma") (region "dend") (region "axon"))', "k", rev_pot=-77)
cell.paint('(join (region "soma") (region "dend") (region "axon"))', hh_na_k_chnl)

cell.paint('"everywhere"', leaky_chnl)
```
Note that `pas` has no `el` parameter in arbor, it is called `e` instead. 

Hope this fixes your issues.

**Reply by lungsi** [reply]
Thank you so much @noraabiakar  (setting the mechanism has certainly changed from the previous version)

BTW In the previous version I had `cell.paint("everywhere", leaky_chnl)`  **how is `cell.paint('"everywhere"', leaky_chnl)` different?**

PS: With the latest version only `cell_model.probe("voltage", '"stim"', 10000)` works **not** `cell_model.probe("voltage", "stim", 10000)`.

**Reply by noraabiakar** [reply]
We've been working on the user facing interface in the past few months. One of the things we've added in v0.4 is the ability to have arbitrary strings as labels for regions/locsets. For example, you can rename your `everywhere` region in the label dictionary to `full cell` or `soma & dend & axon` and that would work, but because there's now spaces and special characters involved, we need to add the quotes to signify that this is a region/locset label as opposed to a region/locset expression.
As an example of why the extra quotes are needed we can look at 2 different methods to paint a `pas` mechanism every where on the cell: 
```
cell.paint('(join (region "soma") (region "dend") (region "axon") (region "apic"))', 'pas')
```
or 
```
# label dictionary defines the "full cell" region
cell.paint('"full cell"', 'pas')
```
If we don't add extra quotes around "full cell" region in the second method, it could be interpreted as a region s-expression/definition. We add the quotes to signify this is instead a region label. It's the same explanation for why we need quotes around `stim`. 

The API changed slightly between releases as we try to settle on an easy interface for the users. We're also working on keeping our documentation up-to-date, the [python examples](https://arbor.readthedocs.io/en/latest/tutorial/single_cell_model.html) are a good place to start to see what has changed since the last release.

**Comment by lungsi** [comment]
## I am afraid this was not the solution. I still get a similar RuntimeError

Starting with the (modified) suggested solution
```
# Set mechanisms
leaky_chnl = arbor.mechanism('passive/e=-54.3', {'g': 0.0003})
hh_na_k_chnl = arbor.mechanism("hh", {"gnabar": 0.120,
                                      "gkbar": 0.036,
                                      "gl": 0.}) # to avoid duplication

# Attach mechanism
cell.paint('(join (region "soma") (region "dend") (region "axon"))', "na", rev_pot=50)
cell.paint('(join (region "soma") (region "dend") (region "axon"))', "k", rev_pot=-77)
cell.paint('(join (region "soma") (region "dend") (region "axon"))', hh_na_k_chnl)
cell.paint('"everywhere"', leaky_chnl)
```
and then `cell_model.run(100)` returns
```
---------------------------------------------------------------------------
RuntimeError                              Traceback (most recent call last)
<ipython-input-6-41323ba3c1b4> in <module>
----> 1 cell_model.run(100) # 100ms

RuntimeError: no mechanism passive in catalogue
```

Also, if I used `leaky_chnl = arbor.mechanism("pass", {'e': -54.3, 'g': 0.0003})` the result is similar
```
---------------------------------------------------------------------------
RuntimeError                              Traceback (most recent call last)
<ipython-input-6-41323ba3c1b4> in <module>
----> 1 cell_model.run(100) # 100ms

RuntimeError: no mechanism pass in catalogue
```
And exactly the same (above) if I use `leaky_chnl = arbor.mechanism('pass/e=-54.3', {'g': 0.0003})`

**Reply by noraabiakar** [reply]
Oh that's my bad, I forgot to change the `passive`/`pas` mechanism in the suggested code, here's what that should look like:
```
# Set mechanisms
leaky_chnl = arbor.mechanism('pas/e=-54.3', {'g': 0.0003})
hh_na_k_chnl = arbor.mechanism("hh", {"gnabar": 0.120,
                                      "gkbar": 0.036,
                                      "gl": 0.}) # to avoid duplication

# Attach mechanism
cell.paint('(join (region "soma") (region "dend") (region "axon"))', "na", rev_pot=50)
cell.paint('(join (region "soma") (region "dend") (region "axon"))', "k", rev_pot=-77)
cell.paint('(join (region "soma") (region "dend") (region "axon"))', hh_na_k_chnl)
cell.paint('"everywhere"', leaky_chnl)
```
I will edit the above answer to use `pas` to help anyone else who may read this discussion.

**Reply by lungsi** [reply]
Ok Thanks @noraabiakar But I get another RuntimeError, this time the error message says `mechanism pas/e` has no parameter called `e`
```
---------------------------------------------------------------------------
RuntimeError                              Traceback (most recent call last)
<ipython-input-6-41323ba3c1b4> in <module>
----> 1 cell_model.run(100) # 100ms

RuntimeError: mechanism pas/e=-54.3 has no parameter e
```

PS: Although you already mentioned above that the `pas` mechanism does not have the `el` parameter I just wanted to give it a try with that and I got a similar RuntimeError, `RuntimeError: mechanism pas/el=-54.3 has no parameter el`

**Comment by lungsi** [comment, marked as answer]
Ok **I think** I got it to work with
```
# Set mechanisms
leaky_chnl = arbor.mechanism("pas", {'e': -54.3, 'g': 0.0003})
hh_na_k_chnl = arbor.mechanism("hh", {"gnabar": 0.120,
                                      "gkbar": 0.036,
                                      "gl": 0.}) # to avoid duplication

# Attach mechanism
cell.paint('(join (region "soma") (region "dend") (region "axon"))', "na", rev_pot=50)
cell.paint('(join (region "soma") (region "dend") (region "axon"))', "k", rev_pot=-77)
cell.paint('(join (region "soma") (region "dend") (region "axon"))', hh_na_k_chnl)
cell.paint('"everywhere"', leaky_chnl)
```
and then `cell_model.run(100)` is successful (i.e, no RuntimeError) but

## I don't know if the simulation runs with the intended changes.

### Note:
- `leaky_chnl = arbor.mechanism('pas/e=-54.3', {'g': 0.0003})` results in RuntimeError

**Reply by noraabiakar** [reply]
Arbor's `pas.mod` mechanism has the following header:
```
NEURON {
    SUFFIX pas
    NONSPECIFIC_CURRENT i
    RANGE g, e
}
```
The `g` and `e` parameters are `RANGE` parameters in this definition, meaning that every instance of this mechanism has its own `g` and `e` values which are controlled/provided by the user, making them inputs to the mechanism. If `g` and `e` had been `GLOBAL` parameters instead, they wouldn't be something to be provided by the user, but more of variables that control the behavior of the mechanism, which makes them a property of the mechanism itself. 

In practice, the distinction between `RANGE` and `GLOBAL` is often murky in NMODL. But the classification is important in determining whether we need to use `arbor.mechanism("pas/e=-54.3")` or `arbor.mechanism("pas", {'e': -54.3})`. 

The first method `arbor.mechanism("pas/e=-54.3")` is used when `e` is a `GLOBAL` parameter: something that controls the behavior of `pas`. Changing `e` changes the intended definition of the mechanism, so we need to `derive` a new mechanism similar to `pas` but with different `GLOBAL` parameters, the resulting mechanism is a new one called `pas/e=-54.3`.

The second method `arbor.mechanism("pas", {'e': -54.3})` is used when `e` is a `RANGE` parameter: something that is meant to be set by the user. Changing `e` doesn't change the mechanism itself, it's like passing arguments to a function. We don't derive a new mechanism in this case, we use the same `pas` mechanism with different inputs. 

So your final solution is indeed correct, and the simulation should run with the intended changes. I hope this explains the runtime errors you have seen. We also have a [page](https://arbor.readthedocs.io/en/latest/concepts/mechanisms.html) in our docs about mechanisms, which may provide more insight.
